R/class-workbook.R

```diff
@@ -5400,6 +5728,281 @@ wbWorkbook <- R6::R6Class(
+              values[[1]],
+              values[[2]],
+              posColour,
+              guid
+            )
+          }
+        },
+
+        expression = {
+          sprintf(
+            '<cfRule type="expression" dxfId="%s" priority="1"><formula>%s</formula></cfRule>',
+            dxfId,
+            formula
+          )
+        },
+
+        duplicatedValues = {
+          sprintf(
+            '<cfRule type="duplicateValues" dxfId="%s" priority="1"/>',
+            dxfId
+          )
+        },
+
+        containsText = {
+          sprintf(
+            '<cfRule type="containsText" dxfId="%s" priority="1" operator="containsText" text="%s">
+                        	<formula>NOT(ISERROR(SEARCH("%s", %s)))</formula>
+                       </cfRule>',
+            dxfId,
+            values,
+            values,
+            # is this unlist correct?  Would this not work?
+            # > strsplit(sqref, split = ":")[[1]]
+            unlist(strsplit(sqref, split = ":"))[[1]]
+          )
+        },
+
+        notContainsText = {
+          sprintf(
+            '<cfRule type="notContainsText" dxfId="%s" priority="1" operator="notContains" text="%s">
+                        	<formula>ISERROR(SEARCH("%s", %s))</formula>
+                       </cfRule>',
+            dxfId,
+            values,
+            values,
+            unlist(strsplit(sqref, split = ":"))[[1]]
+          )
+        },
+
+        beginsWith = {
+          sprintf(
+            '<cfRule type="beginsWith" dxfId="%s" priority="1" operator="beginsWith" text="%s">
+                        	<formula>LEFT(%s,LEN("%s"))="%s"</formula>
+                       </cfRule>',
+            dxfId,
+            values,
+
+            unlist(strsplit(sqref, split = ":"))[[1]],
+            values,
+            values
+          )
+        },
+
+        endsWith = sprintf(
+          '<cfRule type="endsWith" dxfId="%s" priority="1" operator="endsWith" text="%s">
+                        	<formula>RIGHT(%s,LEN("%s"))="%s"</formula>
+                       </cfRule>',
+          dxfId,
+          values,
+
+          unlist(strsplit(sqref, split = ":"))[[1]],
+          values,
+          values
+        ),
+
+        between = sprintf(
+          '<cfRule type="cellIs" dxfId="%s" priority="1" operator="between"><formula>%s</formula><formula>%s</formula></cfRule>',
+          dxfId,
+          formula[1],
+          formula[2]
+        ),
+
+        topN = sprintf(
+          '<cfRule type="top10" dxfId="%s" priority="1" rank="%s" percent="%s"></cfRule>',
+          dxfId,
+          values$rank,
+          values$percent
+        ),
+
+        bottomN = {
+          sprintf(
+            '<cfRule type="top10" dxfId="%s" priority="1" rank="%s" percent="%s" bottom="1"></cfRule>',
+            dxfId,
+            values$rank,
+            values$percent
+          )
+        },
+        # do we have a match.arg() anywhere or will it just be showned in this switch()?
```

Comment:
Either it is shown here or I the stop condition might never be reached. Unsure

Reply:
I had added a `match.arg()` in `$add_conditional_formatting()` so we can change this to an internal error in case something really goes wrong